packages/e2e-tests/plugins/interactive-blocks/directive-run/view.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	store,
+	directive,
+	navigate,
+	useInit,
+	useWatch,
+	cloneElement,
+	getElement,
+} from '@wordpress/interactivity';
+
+// Custom directive to show hide the content elements in which it is placed.
+directive(
+	'show-children',
+	( { directives: { 'show-children': showChildren }, element, evaluate } ) => {
+		const entry = showChildren.find(
+			( { suffix } ) => suffix === 'default'
+		);
+		return evaluate( entry )
+			? element
+			: cloneElement( element, { children: null } );
+	},
+	{ priority: 9 }
+);
+
+const html = `
+<div
+	data-wp-interactive='{ "namespace": "directive-run" }'
+	data-wp-navigation-id='test-directive-run'
+>
+	<div data-testid="hydrated" data-wp-text="state.isHydrated"></div>
+	<div data-testid="mounted" data-wp-text="state.isMounted"></div>
+	<div data-testid="renderCount" data-wp-text="state.renderCount"></div>
+	<div data-testid="navigated">yes</div>
+
+	<div
+		data-wp-run--hydrated="callbacks.updateIsHydrated"
+		data-wp-run--renderCount="callbacks.updateRenderCount"
+		data-wp-text="state.clickCount"
+	></div>
+</div>
+`;
+
+const { state } = store( 'directive-run', {
+	state: {
+		isOpen: false,
+		isHydrated: 'no',
+		isMounted: 'no',
+		renderCount: 0,
+		clickCount: 0
+	},
+	actions: {
+		toggle() {
+			state.isOpen = ! state.isOpen;
+		},
+		increment() {
+			state.clickCount = state.clickCount + 1;
+		},
+		navigate() {
+			navigate( window.location, {
+				force: true,
+				html,
+			} );
+		},
+	},
+	callbacks: {
+		updateIsHydrated() {
+			setTimeout( () => ( state.isHydrated = 'yes' ) );
```

Comment:
What's with the `setTimeout`s here?

Reply:
I know this is weird. I saw an infinite loop issue, at least with `state.renderCount` (line 76), so I eventually decided to use `setTimeout` for all callbacks.

PS: I don't know, but regarding [this comment](https://github.com/WordPress/gutenberg/pull/57805/files#r1450610969), maybe these `callbacks` should have been `runs` (or the other way around 🤔).

Reply:
Shouldn't we review that infinite loop issue?

Reply:
Any luck with this?

Reply:
Oh, wait. Isn't this just a "don't use `setState` inside the body function" kind of infinite loop?

You can't do this:

```js
const Comp = () => {
  const [state, setState] = useState(1);
  setState(state + 1); // infinite loop
};
```

You have to do this, which is equivalent to the `setTimeout`:

```js
const Comp = () => {
  const [state, setState] = useState(1);
  useEffect(() => {
    setState(state + 1); // still bad, but better :D
  });
};
```

Reply:
I haven't had a chance to investigate yet, but probably it is what you mention. I'll check tomorrow. 😄